### DISCUSSION TITLE
Precision of AR-measurements

### DISCUSSION THREAD

**Discussion by AndreasHoffmann2** [streetcomplete/StreetComplete#3832] [Feedback]
I started to use the Android-AR to measure streets and cycleways. But the results vary wildly.
For example, I measured the same cycleway twice. The first time, I did just tilt the phone to tag the other side, the second time I walked over to the other side.
The measurements were 0.9 and 1.4 m.

Is this just my phone? I use a Huawei P20 pro.
Do other people have the same amount of inaccuracies?

**Comment by HolgerJeromin** [comment]
> For example, I measured the same cycleway twice. The first time, I did just tilt the phone to tag the other side, the second time I walked over to the other side.
> The measurements were 0.9 and 1.4 m.

The interesting bit is: What is the true width? 
With that, we could teach people to use that method...

**Reply by AndreasHoffmann2** [reply]
1.5 would probably have been the correct value (including the curbs).
If there is no feedback from others, I will do some more experiments.

**Comment by westnordost** [comment]
That is indeed a big difference. What do you mean you walked over to the other side? How far away were you from the cycleway when measuring it the first time?

In general, the further away you are from the arrow, the more imprecise the result. When measuring such short distances, I did not notice any inaccuracies. But when testing, I also stood directly next to it.
Measuring a 6m wide road by just tilting the phone and not crossing over to the other side resulted in an imprecision of ~3% or 20cm (shorter than in reality) in my tests.

**Reply by AndreasHoffmann2** [reply]
Both times, I stood between 0.5 and 1 m away for the first measurement.
When measuring 0.9, I just tilted the phone. For the 1.4, I walked the few steps onto the cycleway  and measured the other side under the same angle.

Currently, it seems as if this issue isn't too common since nobody else answered here with the same issue.  I'll do some more tests in the next couple of days with the different phones my family owns.

A spontaneous test inside with a 2m yardstick resulted in perfect 2.0 m with both methods.

**Comment by westnordost** [comment]
See also https://github.com/streetcomplete/StreetComplete/discussions/3802#discussioncomment-2247490 for prior tests / feedback on this feature

**Comment by AndreasHoffmann2** [comment]
I made some more tests and focused on reproducing the inconsistent measurements. I used my Huawei P20 pro and the Xiaomi Redmi Note 9. Both showed the same behavior.
My test-setup: Stand still and set the first arrow less than 1m in front of me. Without moving forward, tilt the phone upwards and measure a point about 2 m further away.
Sometimes, the results were pretty accurate and consistent. No matter how I moved the phone, the measurement was good.

But I had pretty inconsistent measurements as well. At one location, the distance kept flickering 2-3 times a second between 1.8 and 2.5 m. Even though I was holding the phone as still as I could.
I took screenshots in another situation: https://www.dropbox.com/sh/cszrzmp2ci68tuf/AAACpADt37g3gSvOnOvXBRLda?dl=0
The 2.0 m are correct, but the 0.9 should be more like 1.7 m.

Can anybody reproduce that kind of behavior?

**Reply by westnordost** [reply]
Interesting. Note the size of the arrow. In one picture, the arrow is not on the ground but floating somewhere above the ground - hence the shorter measured distance. It looks like ARCore on your Android cannot decide whether the sett at the center of your screen is on the ground or if there is something like a "table" at 1m height.

When you tap the screen, the measurement locks in place. Move around the measured line and you'll see that in the "0.9 meters" case, the second point is actually floating in the air.

**Reply by westnordost** [reply]
I implemented now that the second point must be on the same horizontal plane as the first point. So this should solve this particular case.

**Comment by InsertUser** [comment]
I've had some results that seem like they might be a little off, but not _too_ unreasonable. I'd quite like a `source:width=StreetComplete AR` TBH. If something's way off I reject it, but in other cases I'm not sure if it's my estimate that's off.

**Reply by westnordost** [reply]
source:width=ARCore is set

**Reply by InsertUser** [reply]
It seems the way I checked prior to posting was the one where I noticed my second arrow was a few cm short and tweaked the value. My bad.

**Reply by westnordost** [reply]
Yes, if you change the value after measuring with AR, `source:width=ARCore` is not set
